metricflow/query/group_by_item/candidate_push_down/push_down_visitor.py

```diff
@@ -0,0 +1,408 @@
+            )
+            logger.info(
+                f"Handling {node.ui_description} with candidates from parents:\n"
+                + indent_log_line(mf_pformat(merged_result_from_parents.candidate_set.specs))
+            )
+            if merged_result_from_parents.candidate_set.is_empty:
+                return merged_result_from_parents
+
+            metric = self._semantic_manifest_lookup.metric_lookup.get_metric(node.metric_reference)
+
+            # For metrics with offset_to_grain, don't allow date_part group-by-items
+            patterns_to_apply: Sequence[SpecPattern] = ()
+            if metric.type is MetricType.SIMPLE or metric.type is MetricType.CUMULATIVE:
+                pass
+            elif metric.type is MetricType.RATIO or metric.type is MetricType.DERIVED:
+                for input_metric in metric.input_metrics:
+                    if input_metric.offset_to_grain:
+                        # From comment in previous query parser:
+                        # "Cannot extract date part for metrics with offset_to_grain."
+                        patterns_to_apply = (NoneDatePartPattern(),)
+                        break
+            else:
+                assert_values_exhausted(metric.type)
+
+            candidate_specs: Sequence[LinkableInstanceSpec] = merged_result_from_parents.candidate_set.specs
+            issue_sets_to_merge = [merged_result_from_parents.issue_set]
+
+            if len(patterns_to_apply) == 0:
+                return PushDownResult(
+                    candidate_set=GroupByItemCandidateSet(
+                        specs=tuple(candidate_specs),
+                        measure_paths=merged_result_from_parents.candidate_set.measure_paths,
+                        path_from_leaf_node=current_traversal_path,
+                    ),
+                    issue_set=MetricFlowQueryResolutionIssueSet.merge_iterable(issue_sets_to_merge),
+                )
+
+            matched_specs = candidate_specs
+            for pattern_to_apply in patterns_to_apply:
+                matched_specs = InstanceSpecSet.from_specs(pattern_to_apply.match(matched_specs)).linkable_specs
+
+            logger.info(
+                f"For {node.ui_description}:\n"
+                + indent_log_line(
+                    "After applying patterns:\n"
+                    + indent_log_line(mf_pformat(patterns_to_apply))
+                    + "\n"
+                    + "to inputs, outputs are:\n"
+                    + indent_log_line(mf_pformat(matched_specs))
+                )
+            )
+
+            # There were candidates that were common from the ones passed from parents, but after applying the filters,
+            # none of the candidates were valid.
+            if len(matched_specs) == 0:
+                issue_sets_to_merge.append(
+                    MetricFlowQueryResolutionIssueSet.from_issue(
+                        MetricExcludesDatePartIssue.from_parameters(
```

Comment:
Ok, this is a little weird, because all of the intervening logic is general to any arbitrary collection of patterns but this issue is highly specific to date part support.

I don't think we need to fix this now - and it might be tricky to do this cleanly - but it would be nice to align the error handling with the intervening logic. Otherwise, someone will add a pattern and end up with a date part issue even when no date part is in the group by, and that'll be confusing.

Reply:
Yeah, this could be improved, but I couldn't think of a great solution in time. Should be addressed though.